File: 05-Algorithms/code/typescript/src/rabinKarp.ts
/**
 * Rabin-Karp substring search using a fixed deterministic polynomial rolling
 * hash: base=131, modulus=1_000_000_007 (both chosen small enough that
 * intermediate products stay within the safe double-precision integer range
 * for the string lengths this lab exercises). To eliminate false positives
 * from hash collisions entirely -- rather than accepting the usual
 * Monte-Carlo risk -- every hash match is verified with a direct character
 * comparison before being reported, so the returned indices are exact.
 * Returns every start index in `text` where `pat` occurs (overlapping
 * matches included), ascending. O(|text| + |pat|) expected.
 */
const BASE = 131;
const MOD = 1_000_000_007;

export function rabinKarp(text: string, pat: string): number[] {
  const n = text.length;
  const m = pat.length;
  if (m === 0 || m > n) return [];

  let patHash = 0;
  let pow = 1;
  for (let i = 0; i < m; i++) {
    patHash = (patHash * BASE + pat.charCodeAt(i)) % MOD;
    if (i < m - 1) pow = (pow * BASE) % MOD;
  }

  let windowHash = 0;
  for (let i = 0; i < m; i++) {
    windowHash = (windowHash * BASE + text.charCodeAt(i)) % MOD;
  }

  const result: number[] = [];
  for (let i = 0; i + m <= n; i++) {
    if (windowHash === patHash && text.slice(i, i + m) === pat) {
      result.push(i);
    }
    if (i + m < n) {
      windowHash = ((windowHash - text.charCodeAt(i) * pow) % MOD + MOD) % MOD;
      windowHash = (windowHash * BASE + text.charCodeAt(i + m)) % MOD;
    }
  }
  return result;
}

File: 05-Algorithms/code/python/seb_alg/rabin_karp.py
"""Mirrors `typescript/src/rabinKarp.ts` exactly."""

from __future__ import annotations

from typing import List

_BASE = 131
_MOD = 1_000_000_007


def rabin_karp(text: str, pat: str) -> List[int]:
    """Rabin-Karp substring search using a fixed deterministic rolling hash
    (base=131, mod=1_000_000_007); every hash match is verified with a
    direct character comparison, eliminating false positives entirely.
    Every start index of pat in text, ascending, overlaps included.
    O(|text| + |pat|) expected.
    """
    n, m = len(text), len(pat)
    if m == 0 or m > n:
        return []

    pat_hash = 0
    pow_ = 1
    for i in range(m):
        pat_hash = (pat_hash * _BASE + ord(pat[i])) % _MOD
        if i < m - 1:
            pow_ = (pow_ * _BASE) % _MOD

    window_hash = 0
    for i in range(m):
        window_hash = (window_hash * _BASE + ord(text[i])) % _MOD

    result: List[int] = []
    for i in range(n - m + 1):
        if window_hash == pat_hash and text[i : i + m] == pat:
            result.append(i)
        if i + m < n:
            window_hash = (window_hash - ord(text[i]) * pow_) % _MOD
            window_hash = (window_hash + _MOD) % _MOD
            window_hash = (window_hash * _BASE + ord(text[i + m])) % _MOD
    return result
